### PR TITLE
added defer method link and link property

### PR DESCRIPTION
**Pull request by robinnedergard**

### CODE REVIEW

File: Payment/Defer/index.ts [hunk 1]
```diff
@@ -4,11 +4,17 @@ import { Method as DeferMethod } from "./Method"
...
 export interface Defer extends Base, Omit<DeferCreatable, "currency"> {
 	type: "defer"
+	link?: string
 }
 
 export namespace Defer {
 	export function is(value: any | Defer): value is Defer {
-		return DeferCreatable.is(value) && Base.is(value)
+		return (
+			((value.method != "link" && (value.link == undefined || typeof value.link == "string")) ||
```
Comment (simonmika) [inline Payment/Defer/index.ts:13]: The `link` property is unrelated to which method is used. It should always be present except for old Orders.

**Review by simonmika — CHANGES_REQUESTED** [review]

File: Payment/Defer/Creatable.ts [hunk 2]
```diff
@@ -14,7 +14,8 @@ export namespace Creatable {
 			typeof value == "object" &&
 			value.type == "defer" &&
 			Method.is(value.method) &&
-			typeof value.contact == "string" &&
+			((value.contact == undefined && value.method == "link") ||
+				(typeof value.contact == "string" && value.method != "link")) &&
```
Comment (simonmika) [inline Payment/Defer/Creatable.ts:18]: What about:
```
value.method == "link" ? value.contact == undefined : typeof value.contact == "string"
```

File: Payment/Defer/Method.ts [hunk 3]
```diff
@@ -1,7 +1,8 @@
-export type Method = "sms" | "email"
+export type Method = "sms" | "email" | "link"
+const types: Method[] = ["sms", "email", "link"]
```
Comment (simonmika) [inline Payment/Defer/Method.ts:2]: Should be in namespace.

Comment (simonmika) [reply Payment/Defer/Method.ts:2]: And exported.

File: Payment/Defer/index.ts [hunk 4]
```diff
@@ -4,11 +4,12 @@ import { Method as DeferMethod } from "./Method"
 
 export interface Defer extends Base, Omit<DeferCreatable, "currency"> {
 	type: "defer"
+	link?: string
 }
 
 export namespace Defer {
 	export function is(value: any | Defer): value is Defer {
-		return DeferCreatable.is(value) && Base.is(value)
+		return (value.link == undefined || typeof value.link == "string") && DeferCreatable.is(value) && Base.is(value)
```
Comment (simonmika) [inline Payment/Defer/index.ts:12]: Wrong order. Must verify that value is an object first or we will have null reference issues.

**Review by simonmika — CHANGES_REQUESTED** [review]

**Review by simonmika — APPROVED** [review]

**Review by fredrikbryntesson — APPROVED** [review]

**Pull request merged**